components/layout_2020/flexbox/layout.rs

```diff
@@ -2,32 +2,1090 @@
+            // layout of the content to be fully done:
+            // https://github.com/w3c/csswg-drafts/issues/4905
+            // Gecko reportedly uses `block-size: fit-content` in this case
+            // (which requires running another pass of the "full" layout algorithm)
+            todo!()
+            // Note: this panic shouldn’t happen since the start of `FlexContainer::layout`
+            // forces `FlexAxis::Row`.
+        },
+    };
+
+    // “Resolve the flexible lengths of all the flex items to find their *used main size*.”
+    // https://drafts.csswg.org/css-flexbox/#algo-flex
+    let flex_lines = collect_flex_lines(
+        &mut flex_context,
+        container_main_size,
+        &mut flex_items,
+        |flex_context, mut line| line.layout(flex_context, container_main_size),
+    );
+
+    // https://drafts.csswg.org/css-flexbox/#algo-cross-container
+    let container_cross_size = flex_context
+        .container_definite_inner_size
+        .cross
+        .unwrap_or_else(|| {
+            flex_lines
+                .iter()
+                .map(|line| line.cross_size)
+                .sum::<Length>()
+        })
+        .clamp_between_extremums(
+            flex_context.container_min_cross_size,
+            flex_context.container_max_cross_size,
+        );
+
+    // https://drafts.csswg.org/css-flexbox/#algo-line-align
+    let mut cross_start_position_cursor = Length::zero();
+    let line_cross_start_positions = flex_lines
+        .iter()
+        .map(|line| {
+            // FIXME: “Align all flex lines per `align-content`.”
+            // For now we hard-code the behavior of `align-content: flex-start`.
+            let cross_start = cross_start_position_cursor;
+            let cross_end = cross_start + line.cross_size;
+            cross_start_position_cursor = cross_end;
+            cross_start
+        })
+        .collect::<Vec<_>>();
+
+    let content_block_size = match flex_context.flex_axis {
+        FlexAxis::Row => {
+            // `container_main_size` ends up unused here but in this case that’s fine
+            // since it was already excatly the one decided by the outer formatting context.
+            container_cross_size
+        },
+        FlexAxis::Column => {
+            // FIXME: `container_cross_size` ends up unused here, which is a bug.
+            // It is meant to be the used inline-size, but the parent formatting context
+            // has already decided a possibly-different used inline-size.
+            // The spec is missing something to resolve this conflict:
+            // https://github.com/w3c/csswg-drafts/issues/5190
+            // And we’ll need to change the signature of `IndependentFormattingContext::layout`
+            // to allow the inner formatting context to “negociate” a used inline-size
+            // with the outer one somehow.
+            container_main_size
+        },
+    };
+    let fragments = flex_lines
+        .into_iter()
```

Comment:
Can this loop be merged with the previous one?

Reply:
I guess we're typically iterating over a small number of items so it won't matter much

Reply:
This PR was getting big, so at some point I stopped refactoring and focused on getting it landed. I’m sure it can be improved still.

There’s a limitation though where the returned `impl Iterator` needs to take ownership of every closure captures, it cannot borrow stuff on the stack. Some `.collect()` calls help with that.